cli/tests/program.rs

```diff
@@ -1350,6 +1356,210 @@ fn test_cli_program_mismatch_buffer_authority() {
     process_command(&config).unwrap();
 }
 
+#[test]
+fn test_cli_program_deploy_with_offline_signing() {
+    solana_logger::setup();
+
+    let mut noop_path = PathBuf::from(env!("CARGO_MANIFEST_DIR"));
+    noop_path.push("tests");
+    noop_path.push("fixtures");
+    noop_path.push("noop");
+    noop_path.set_extension("so");
+
+    let mut noop_large_path = PathBuf::from(env!("CARGO_MANIFEST_DIR"));
+    noop_large_path.push("tests");
+    noop_large_path.push("fixtures");
+    noop_large_path.push("noop_large");
+    noop_large_path.set_extension("so");
+
+    let mint_keypair = Keypair::new();
+    let mint_pubkey = mint_keypair.pubkey();
+    let faucet_addr = run_local_faucet(mint_keypair, None);
+    let test_validator =
+        TestValidator::with_no_fees(mint_pubkey, Some(faucet_addr), SocketAddrSpace::Unspecified);
+
+    let rpc_client =
+        RpcClient::new_with_commitment(test_validator.rpc_url(), CommitmentConfig::processed());
+
+    let blockhash = rpc_client.get_latest_blockhash().unwrap();
+
+    let mut file = File::open(noop_large_path.to_str().unwrap()).unwrap();
+    let mut large_program_data = Vec::new();
+    file.read_to_end(&mut large_program_data).unwrap();
+    let max_program_data_len = large_program_data.len();
+    let minimum_balance_for_large_buffer = rpc_client
+        .get_minimum_balance_for_rent_exemption(UpgradeableLoaderState::size_of_programdata(
+            max_program_data_len,
+        ))
+        .unwrap();
+
+    let mut config = CliConfig::recent_for_tests();
+    config.json_rpc_url = test_validator.rpc_url();
+
+    let online_signer = Keypair::new();
+    let online_signer_identity = NullSigner::new(&online_signer.pubkey());
+    let offline_signer = Keypair::new();
+    let buffer_signer = Keypair::new();
+    let buffer_signer_identity = NullSigner::new(&buffer_signer.pubkey());
+    // Typically, keypair for program signer should be different from online signer or
+    // offline signer keypairs.
+    let program_signer = Keypair::new();
+    let program_signer_identity = NullSigner::new(&program_signer.pubkey());
+
+    // Assume fee payer will be either online signer or offline signer (could be completely
+    // separate signer too, but that option is unlikely to be chosen often, so don't bother
+    // testing for it), we want to test both
+    let use_offline_signer_as_fee_payer = rand::thread_rng().gen_bool(0.5);
```

Comment:
Random tests are typically a big headache for CI maintainers and an antipattern, since they make tests look flaky when they only fail sometimes.

Let's remove the random dependency -- if we want to test both cases, let's test both! You can use the `test_case` macro for generating multiple tests

Reply:
What I'm doing here with randomness is sort of a short-cut code-writers might sometimes take to get higher code coverage, I wouldn't say it's related to "flaky code" - it's more like Fuzz testing. Flaky typically means "false positive test fail", while for what I'm doing here test failure (with either of: online or offline fee payer, chosen randomly) would mean we have a bug somewhere in our implementation that needs fixing.

I could see though how this approach could complicate debugging test failures, let me see what I can do about it.

Reply:
I meant more that it could make the test *look* flaky, since if one case gets broken in the future, the test will only catch it 50% of the time.

Definitely look into the `test_case` macro: https://docs.rs/test-case/latest/test_case/ we use it all over the code, and it's a great help

Reply:
Thanks for  `test_case` does seems like a good fit here, the only thing is - I had to stick it under `mod tests` to make it compile: https://github.com/solana-labs/solana/pull/33860/commits/f634c953e7d7ebf10ea50bf9f4453ac725b7885c (moved all tests from `cli/tests/program.rs` file under `mod tests`, for consistency),

this is sort off new stuff for me, not sure whether "it's the proper way to do it" - so calling it out for you to check,

the alternative would be to just manually write a `for` loop in `test_cli_program_deploy_with_offline_signing` I guess.

Reply:
The diff is pretty ugly though, perhaps you'd want me to move this change into a separate PR as "a preliminary refactoring" (assuming it's the way to go).

Reply:
Ok cool! But it's strange, I was able to get it to work without moving everything under `mod tests` -- what error did you see without it?

Reply:
Looks like `clippy` isn't happy about it though, it is happy on `+nightly` (https://github.com/rust-lang/rust-clippy/issues/11153 issue has been addressed) - but CI runs older version and fails https://github.com/solana-labs/solana/actions/runs/7100774373/job/19327561856#step:5:2888

separate commit should address it (approach copied from `solana/core/tests/epoch_accounts_hash.rs` file): https://github.com/solana-labs/solana/pull/33860/commits/372b86f40f14bf76f987db3751ecf081d5537a5e

Reply:
Accidentally added some older changes to those commits mentioned  above, now cleaned up (and updated commits referenced in discussions  accordingly).